[BEAM-8376] Google Cloud Firestore Connector - Add Firestore V1 Write Operations

BenWhitehead opened the pull request:
First PR in a series of PRs to add sources and sinks for Google Cloud Firestore.

This PR adds a sink for writing Documents to Cloud Firestore. This PR also 
includes common code that is shared between the sources and Firestore SDK 
adapter which will be added in follow up PRs. (If desired I can split the common
code into a separate PR, but I thought it wouldn't make much sense without the 
context of it being used).

Each commit has a detailed commit message, however a brief summary of the 4 
commits included in this PR are as follows:

### Gradle 
Add Google Cloud Firestore dependencies. Version 2.2.4+ is needed for all dependencies
which include bugfixes and the necessary grpc client, protos and bootstraping
code.

### Common Classes
A set of base classes which underpin the added PTransforms and DoFns that will
make up the connector. This change also includes the base classes used in the
unit and integration test suites which are present for each component of the
connector.

### RPC QoS Classes
All read and write RPC Operations are managed in coordination with the RPC QoS
layer. Attempt counts, backoff amounts, batch sizing, success, failure metrics,
throttling are all managed here.

### Firestore v1 Write Operations
Add PTransforms for writing to Cloud Firestore via the BatchWrite API. There are
two implementations which only differ in what happens when an error is 
encountered. The first implementation will throw an exception if an error is 
encountered, whereas the second implementation will output a failure object 
(functioning similar to a dead letter queue).

## Integration Tests
The integration test suite will only run if `GOOGLE_APPLICATION_CREDENTIALS` or
`FIRESTORE_EMULATOR_HOST` are present in the environment. The project the tests
run against must have a [Firestore in Native mode](https://cloud.google.com/datastore/docs/firestore-or-datastore#in_native_mode)
database configured (Firestore in Native Mode and Firestore in Datastore Mode cannot
both be present in the same project at this time). All integration tests are self
contained and do not depend on preconfigured external state, and clean up after
themselves.

A future PR will be created for sources, and another PR for a Firestore SDK
adapter which both will leverage the common code added in this PR.

## Reviewers

R: @chamikaramj
R: @jlara310 (Firestore Technical Writer)
R: @clement (Firestore Dataplane TL)
R: @danthev (Firestore SRE)

------------------------

Thank you for your contribution! Follow this checklist to help us incorporate your contribution quickly and easily:

 - [ ] [**Choose reviewer(s)**](https://beam.apache.org/contribute/#make-your-change) and mention them in a comment (`R: @username`).
 - [ ] Format the pull request title like `[BEAM-XXX] Fixes bug in ApproximateQuantiles`, where you replace `BEAM-XXX` with the appropriate JIRA issue, if applicable. This will automatically link the pull request to the issue.
 - [ ] Update `CHANGES.md` with noteworthy changes.
 - [ ] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

See the [Contributor Guide](https://beam.apache.org/contribute) for more tips on [how to make review process smoother](https://beam.apache.org/contribute/#make-reviewers-job-easier).

Post-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

Lang | SDK | Dataflow | Flink | Samza | Spark | Twister2
--- | --- | --- | --- | --- | --- | ---
Go | [![Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Go/lastCompletedBuild/badge/icon)](https://ci-beam.apache.org/job/beam_PostCommit_Go/lastCompletedBuild/) | --- | [![Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Go_VR_Flink/lastCompletedBuild/badge/icon)](https://ci-beam.apache.org/job/beam_PostCommit_Go_VR_Flink/lastCompletedBuild/) | --- | [![Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Go_VR_Spark/lastCompletedBuild/badge/icon)](https://ci-beam.apache.org/job/beam_PostCommit_Go_VR_Spark/lastCompletedBuild/) | ---
Java | [![Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java/lastCompletedBuild/badge/icon)](https://ci-beam.apache.org/job/beam_PostCommit_Java/lastCompletedBuild/) | [![Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow/lastCompletedBuild/badge/icon)](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow/lastCompletedBuild/)<br>[![Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_VR_Dataflow_V2/lastCompletedBuild/badge/icon)](https://ci-beam.apache.org/job/beam_PostCommit_Java_VR_Dataflow_V2/lastCompletedBuild/)<br>[![Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow_Java11/lastCompletedBuild/badge/icon)](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow_Java11/lastCompletedBuild/) | [![Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Flink/lastCompletedBuild/badge/icon)](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Flink/lastCompletedBuild/)<br>[![Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Flink_Java11/lastCompletedBuild/badge/icon)](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Flink_Java11/lastCompletedBuild/)<br>[![Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_PVR_Flink_Batch/lastCompletedBuild/badge/icon)](https://ci-beam.apache.org/job/beam_PostCommit_Java_PVR_Flink_Batch/lastCompletedBuild/)<br>[![Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_PVR_Flink_Streaming/lastCompletedBuild/badge/icon)](https://ci-beam.apache.org/job/beam_PostCommit_Java_PVR_Flink_Streaming/lastCompletedBuild/) | [![Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Samza/lastCompletedBuild/badge/icon)](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Samza/lastCompletedBuild/) | [![Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Spark/lastCompletedBuild/badge/icon)](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Spark/lastCompletedBuild/)<br>[![Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_PVR_Spark_Batch/lastCompletedBuild/badge/icon)](https://ci-beam.apache.org/job/beam_PostCommit_Java_PVR_Spark_Batch/lastCompletedBuild/)<br>[![Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_SparkStructuredStreaming/lastCompletedBuild/badge/icon)](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_SparkStructuredStreaming/lastCompletedBuild/) | [![Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Twister2/lastCompletedBuild/badge/icon)](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Twister2/lastCompletedBuild/)
Python | [![Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Python36/lastCompletedBuild/badge/icon)](https://ci-beam.apache.org/job/beam_PostCommit_Python36/lastCompletedBuild/)<br>[![Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Python37/lastCompletedBuild/badge/icon)](https://ci-beam.apache.org/job/beam_PostCommit_Python37/lastCompletedBuild/)<br>[![Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Python38/lastCompletedBuild/badge/icon)](https://ci-beam.apache.org/job/beam_PostCommit_Python38/lastCompletedBuild/) | [![Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Py_VR_Dataflow/lastCompletedBuild/badge/icon)](https://ci-beam.apache.org/job/beam_PostCommit_Py_VR_Dataflow/lastCompletedBuild/)<br>[![Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Py_VR_Dataflow_V2/lastCompletedBuild/badge/icon)](https://ci-beam.apache.org/job/beam_PostCommit_Py_VR_Dataflow_V2/lastCompletedBuild/)<br>[![Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Py_ValCont/lastCompletedBuild/badge/icon)](https://ci-beam.apache.org/job/beam_PostCommit_Py_ValCont/lastCompletedBuild/) | [![Build Status](https://ci-beam.apache.org/job/beam_PreCommit_Python_PVR_Flink_Cron/lastCompletedBuild/badge/icon)](https://ci-beam.apache.org/job/beam_PreCommit_Python_PVR_Flink_Cron/lastCompletedBuild/)<br>[![Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Python_VR_Flink/lastCompletedBuild/badge/icon)](https://ci-beam.apache.org/job/beam_PostCommit_Python_VR_Flink/lastCompletedBuild/) | --- | [![Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Python_VR_Spark/lastCompletedBuild/badge/icon)](https://ci-beam.apache.org/job/beam_PostCommit_Python_VR_Spark/lastCompletedBuild/) | ---
XLang | [![Build Status](https://ci-beam.apache.org/job/beam_PostCommit_XVR_Direct/lastCompletedBuild/badge/icon)](https://ci-beam.apache.org/job/beam_PostCommit_XVR_Direct/lastCompletedBuild/) | [![Build Status](https://ci-beam.apache.org/job/beam_PostCommit_XVR_Dataflow/lastCompletedBuild/badge/icon)](https://ci-beam.apache.org/job/beam_PostCommit_XVR_Dataflow/lastCompletedBuild/) | [![Build Status](https://ci-beam.apache.org/job/beam_PostCommit_XVR_Flink/lastCompletedBuild/badge/icon)](https://ci-beam.apache.org/job/beam_PostCommit_XVR_Flink/lastCompletedBuild/) | --- | [![Build Status](https://ci-beam.apache.org/job/beam_PostCommit_XVR_Spark/lastCompletedBuild/badge/icon)](https://ci-beam.apache.org/job/beam_PostCommit_XVR_Spark/lastCompletedBuild/) | ---

Pre-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

--- |Java | Python | Go | Website | Whitespace | Typescript
--- | --- | --- | --- | --- | --- | ---
Non-portable | [![Build Status](https://ci-beam.apache.org/job/beam_PreCommit_Java_Cron/lastCompletedBuild/badge/icon)](https://ci-beam.apache.org/job/beam_PreCommit_Java_Cron/lastCompletedBuild/) | [![Build Status](https://ci-beam.apache.org/job/beam_PreCommit_Python_Cron/lastCompletedBuild/badge/icon)](https://ci-beam.apache.org/job/beam_PreCommit_Python_Cron/lastCompletedBuild/)<br>[![Build Status](https://ci-beam.apache.org/job/beam_PreCommit_PythonLint_Cron/lastCompletedBuild/badge/icon)](https://ci-beam.apache.org/job/beam_PreCommit_PythonLint_Cron/lastCompletedBuild/)<br>[![Build Status](https://ci-beam.apache.org/job/beam_PreCommit_PythonDocker_Cron/badge/icon)](https://ci-beam.apache.org/job/beam_PreCommit_PythonDocker_Cron/lastCompletedBuild/) <br>[![Build Status](https://ci-beam.apache.org/job/beam_PreCommit_PythonDocs_Cron/badge/icon)](https://ci-beam.apache.org/job/beam_PreCommit_PythonDocs_Cron/lastCompletedBuild/) | [![Build Status](https://ci-beam.apache.org/job/beam_PreCommit_Go_Cron/lastCompletedBuild/badge/icon)](https://ci-beam.apache.org/job/beam_PreCommit_Go_Cron/lastCompletedBuild/) | [![Build Status](https://ci-beam.apache.org/job/beam_PreCommit_Website_Cron/lastCompletedBuild/badge/icon)](https://ci-beam.apache.org/job/beam_PreCommit_Website_Cron/lastCompletedBuild/) | [![Build Status](https://ci-beam.apache.org/job/beam_PreCommit_Whitespace_Cron/lastCompletedBuild/badge/icon)](https://ci-beam.apache.org/job/beam_PreCommit_Whitespace_Cron/lastCompletedBuild/) | [![Build Status](https://ci-beam.apache.org/job/beam_PreCommit_Typescript_Cron/lastCompletedBuild/badge/icon)](https://ci-beam.apache.org/job/beam_PreCommit_Typescript_Cron/lastCompletedBuild/)
Portable | --- | [![Build Status](https://ci-beam.apache.org/job/beam_PreCommit_Portable_Python_Cron/lastCompletedBuild/badge/icon)](https://ci-beam.apache.org/job/beam_PreCommit_Portable_Python_Cron/lastCompletedBuild/) | --- | --- | --- | ---

See [.test-infra/jenkins/README](https://github.com/apache/beam/blob/master/.test-infra/jenkins/README.md) for trigger phrase, status and link of all Jenkins jobs.


GitHub Actions Tests Status (on master branch)
------------------------------------------------------------------------------------------------
[![Build python source distribution and wheels](https://github.com/apache/beam/workflows/Build%20python%20source%20distribution%20and%20wheels/badge.svg?branch=master&event=schedule)](https://github.com/apache/beam/actions?query=workflow%3A%22Build+python+source+distribution+and+wheels%22+branch%3Amaster+event%3Aschedule)
[![Python tests](https://github.com/apache/beam/workflows/Python%20tests/badge.svg?branch=master&event=schedule)](https://github.com/apache/beam/actions?query=workflow%3A%22Python+Tests%22+branch%3Amaster+event%3Aschedule)
[![Java tests](https://github.com/apache/beam/workflows/Java%20Tests/badge.svg?branch=master&event=schedule)](https://github.com/apache/beam/actions?query=workflow%3A%22Java+Tests%22+branch%3Amaster+event%3Aschedule)

See [CI.md](https://github.com/apache/beam/blob/master/CI.md) for more information about GitHub Actions CI.

danthev commented on sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/firestore/RpcQosImpl.java:
```diff
@@ -0,0 +1,695 @@
+  @VisibleForTesting
+  static final class WriteRampUp {
+    private static final Duration RAMP_UP_INTERVAL = Duration.standardMinutes(5);
+    private final int baseMax;
```
`initialWriteBudget` or `baseBatchBudget`?

BenWhitehead replied:
I'll change it to `baseBatchBudget`

danthev commented on sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/firestore/RpcQosOptions.java:
```diff
@@ -0,0 +1,666 @@
+import static java.util.Objects.requireNonNull;
+import static org.apache.beam.vendor.guava.v26_0_jre.com.google.common.base.Preconditions.checkArgument;
+
+import edu.umd.cs.findbugs.annotations.Nullable;
```
Wrong import.

danthev commented on sdks/java/io/google-cloud-platform/src/test/java/org/apache/beam/sdk/io/gcp/firestore/BaseFirestoreV1WriteFnTest.java:
```diff
@@ -0,0 +1,640 @@
+  @Test
+  public final void attemptsExhaustedForRetryableError() throws Exception {
+    Instant attemptStart = Instant.ofEpochMilli(0);
+    Instant rpc1Sta = Instant.ofEpochMilli(1);
```
I think we can add two more letters to make it `rpc1Start`. :)

BenWhitehead replied:
Will update

danthev commented on sdks/java/io/google-cloud-platform/src/test/java/org/apache/beam/sdk/io/gcp/firestore/BaseFirestoreV1WriteFnTest.java:
```diff
@@ -0,0 +1,640 @@
+
+    Write write = FirestoreProtoHelpers.newWrite();
+    BatchWriteRequest expectedRequest = BatchWriteRequest.newBuilder()
+        .setDatabase("projects/testing-project/databases/(default)")
```
Does this need to be parametrized on `ENV_GOOGLE_CLOUD_PROJECT`? Same on the other test cases.

BenWhitehead replied:
not in this case, the unit tests are all in process and use mocks for all RPCs.

danthev replied:
Ok.

danthev reviewed (changes requested):
Is there a way to preview the generated Java Doc? I only scanned most of the doc strings.

BenWhitehead commented on sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/firestore/FirestoreDoFn.java:
```diff
@@ -0,0 +1,85 @@
+  @StartBundle
+  public abstract void startBundle(DoFn<In, Out>.StartBundleContext context) throws Exception;
+
+  static abstract class NonWindowAwareDoFn<In, Out> extends FirestoreDoFn<In, Out> {
```
Yup, it's used by most of the Read operations. I missed removing this when I was splitting things up. I'll remove this for now.

BenWhitehead commented on sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/firestore/FirestoreIOOptions.java:
```diff
@@ -0,0 +1,49 @@
+   * @param hostPort the emulator host and port to connect to
+   * @see FirestoreOptions.Builder#setEmulatorHost(java.lang.String)
+   */
+  void setEmulatorHostPort(String hostPort);
```
Sure, I'll update

BenWhitehead commented on sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/firestore/FirestoreStatefulComponentFactory.java:
```diff
@@ -0,0 +1,129 @@
+          @Override
+          public Map<@NonNull String, @NonNull String> getHeaders() {
+            String version = "x.y.z";  // TODO: How does beam manage version detection in project?
+            return ImmutableMap.of("User-Agent", String.format("beam-sdk/%s", version));
```
I'll update to user that

BenWhitehead commented on sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/firestore/FirestoreV1WriteFn.java:
```diff
@@ -0,0 +1,526 @@
+     */
+    @Override
+    public void processElement(ProcessContext context, BoundedWindow window) throws Exception {
+      @SuppressWarnings("nullness") // for some reason requireNonNull thinks its parameter must be non-null...
```
I spent a little time digging and it appears to be due to the fact that error checker is configured to assume non null unless explicitly annotated, so `Objects.requireNotNull` is being treated as only accepting NonNull arguments. I'll update the comment to state this instead of what's there right now.

BenWhitehead commented on sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/firestore/RpcQosImpl.java:
```diff
@@ -0,0 +1,695 @@
+          new MovingAverage(
+              options.getSamplePeriod(),
+              options.getSamplePeriodBucketSize()
+              /* numSignificantBuckets */
```
Yup, left behind by a refactor. I'll clean up.

BenWhitehead commented on sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/firestore/RpcQosImpl.java:
```diff
@@ -0,0 +1,695 @@
+    public void recordStartRequest(Instant instantSinceEpoch) {
+      at.recordStartRequest(instantSinceEpoch);
+      start = instantSinceEpoch;
+      state = AttemptState.Active_Started;
```
Will do

BenWhitehead commented on sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/firestore/FirestoreV1WriteFn.java:
```diff
@@ -0,0 +1,526 @@
+      int writesCount = flushBuffer.getBufferedElementsCount();
+      long bytes = flushBuffer.getBufferedElementsBytes();
+      BatchWriteRequest request = getBatchWriteRequest(flushBuffer);
+      while (true) {
```
Will do

BenWhitehead commented on sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/firestore/FirestoreV1WriteFn.java:
```diff
@@ -0,0 +1,526 @@
+     * happen and how many writes are included is determined and managed by the {@link RpcQos}
+     * instance of this class.
+     *
+     * @param finalFlush A boolean specifying if this call is from {@link #finishBundle(DoFn.FinishBundleContext)}. If
```
I've renamed the parameter to `finishingBundle` to try and more clearly communicate that system state associated with the property and why the code might care. I've also elaborated on the description for the case when false is passed. Does this get to what you were hoping to see?

danthev replied:
What I was trying to say is that, semantically, I would describe in the Javadoc what a parameter does, with the "why" as 
a sidenote. So the added second sentence is good, and with the rename it matches what's in the first sentence.

BenWhitehead commented:
Thanks for the review @danthev!

JavaDocs can be generated locally by running the following from you projects base dir:
```
./gradlew javadoc
open $(pwd)/sdks/java/io/google-cloud-platform/build/docs/javadoc/index.html
```

BenWhitehead commented on sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/firestore/FirestoreV1.java:
```diff
@@ -0,0 +1,610 @@
+     * FirestoreIO#v1()}.
+     * <p/>
+     *
+     * <b>Example Usage</b>
```
Done

danthev reviewed (commented):
Javadoc LGTM, you'll just need to add a [package-info.java](https://github.com/apache/beam/blob/38c9caa0dcf8cec58dfaee18547d9151fd500f44/sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/datastore/package-info.java) to get a Javadoc for the package generally.

jlara310 commented on sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/firestore/RpcQos.java:
```diff
@@ -0,0 +1,253 @@
+    void checkCanRetry(RuntimeException exception) throws InterruptedException;
+
+    /**
+     * Record the time start time of sending the rpc
```
"time start time" -> "start time"

jlara310 commented on sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/firestore/RpcQos.java:
```diff
@@ -0,0 +1,253 @@
+
+    /**
+     * An element which can be added to a {@link FlushBuffer}. This interface is mainly a marker for
+     * ensuring an encapsulated lifecycle managed way of determining the serialized size of from
```
"of from" -> "of" ?

jlara310 commented on sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/firestore/RpcQosImpl.java:
```diff
@@ -0,0 +1,695 @@
+   * The target minimum number of requests per samplePeriodMs, even if no requests succeed. Must be
+   * greater than 0, else we could throttle to zero. Because every decision is probabilistic, there
+   * is no guarantee that the request rate in any given interval will not be zero. (This is the +1
+   * from the formula in https://landing.google.com/sre/book/chapters/handling-overload.html
```
Add a closing parenthesis.

jlara310 commented on sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/firestore/RpcQosImpl.java:
```diff
@@ -0,0 +1,695 @@
+   * Determines batch sizes for commit RPCs based on past performance.
+   *
+   * <p>It aims for a target response time per RPC: it uses the response times for previous RPCs and
+   * the number of entities contained in them, calculates a rolling average time-per-document, and
```
Here and in the next line: "entities" -> "documents"

jlara310 reviewed (commented):
Reviewed Javadoc comments for readability. LGTM after addressing review comments.

chamikaramj commented:
Run Java PreCommit

chamikaramj commented:
Run Spotless PreCommit

chamikaramj commented on sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/firestore/FirestoreDoFn.java:
```diff
@@ -0,0 +1,66 @@
+import org.apache.beam.sdk.transforms.windowing.BoundedWindow;
+
+/**
+ * Base class for all {@link DoFn} defined in the Firestore Connector.
```
Are you sure this will be the base class for all DoFns used by source and sink ? A DoFn can include any generic functionality and many sources/sinks break up functionality into multiple DoFns due to various reasons (readability, modularity, persistence of input, etc.)

BenWhitehead replied:
It is true for all of the `DoFn` which are stateful and have a lifecycle that must be managed. For the `DoFn` which perform RPCs, Setup is for the RpcQos management which we want to live longer than an individual bundle, and StartBundle is where we create the RPC clients based on the `PipelineOptions` from `StartBundleContext#getPipelineOptions()`.

This base class also serves as an upper bound for the unit tests where all DoFn are checked to ensure they are serializable as well as performing some common mocking which is used across the majority of tests (e.g. `StartBundleContext`, `ProcessContext`, `Credentials` etc).

There are a few `DoFn` which will be added in the "Read" PR to follow which directly extend DoFn and only implement a `@ProcessElement` method as they are otherwise stateless.

I'll update the comment on this class to call out that it is specifically for Stateful `DoFn`.

chamikaramj commented on sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/firestore/FirestoreDoFn.java:
```diff
@@ -0,0 +1,66 @@
+   * @see org.apache.beam.sdk.transforms.DoFn.StartBundle
+   */
+  @StartBundle
+  public abstract void startBundle(DoFn<In, Out>.StartBundleContext context) throws Exception;
```
I think it will be fine to just remove FirestoreDoFn and add these annotations to actual DoFn implementations.

chamikaramj commented on sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/firestore/FirestoreIOOptions.java:
```diff
@@ -0,0 +1,49 @@
+import org.checkerframework.checker.nullness.qual.Nullable;
+
+@Description("Options used to configure Cloud Firestore IO")
+public interface FirestoreIOOptions extends PipelineOptions {
```
Just 'FirestoreOptions'

BenWhitehead replied:
Fixes

chamikaramj commented on sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/firestore/FirestoreV1.java:
```diff
@@ -0,0 +1,620 @@
+ * The default behavior is to fail a bundle if any single write fails with a non-retryable error.
+ * <pre>{@code
+ * PCollection<Write> writes = ...;
+ * PDone sink = writes
```
Returning PDone from a sink is an anti-pattern since users will not be able to continue the pipeline (for example, write to system X after writing to Firestore). We should think of some useful PCollection that we can return from the Firestore sync. If nothing else, this can just be 'null' values (per-window) or the original input PCollection.

BenWhitehead replied:
This is in addition to providing an optional dead letter queue via `.withDeadLetterQueue()`? Is there a good example to follow for something like this? I got the `PDone` from the datastore connector.

chamikaramj replied:
I think still returning DoFn (even for the general case) is an anti-pattern. Some sources (including Datastore) currently do this but we'd like to change such sources. See following thread.
https://lists.apache.org/thread.html/r1fa3599ef9c0034d2bfa0614012bc864ce7b296228d8cf804c931065%40%3Cuser.beam.apache.org%3E

BenWhitehead replied:
I've read through both of the threads you linked and I'm leaning toward following the `PCollection<TypeDescribingWhatIsComplete>` from the threads.

I would Refactor `BatchWrite` to return a `PCollection<NewWriteResultType>` instead of `PDone`.

Two different things jump to mind for the `NewWriteResultType` and I'm curious if you think I should pick one over the other?

1. `WriteSuccessSummary` which would contain a simple summary of the `BatchWrite` including the number of writes, and the number of bytes.
   ```java
   public final class WriteSuccessSummary {
    private final int numWrites;
    private final long numBytes;
   }
   ```
2. `WriteSuccess` which would be emitted for each processed Write and include the `Write`, and the `WriteResult` and `Status` from Firestore
   ```java
   public static final class WriteFailure implements Serializable {
     private final Write write;
     private final WriteResult writeResult;
     private final Status status;
   }
   ```

I'm personally leaning toward number 1 as it follows the SQL pattern and reduces the amount of data in the pipeline reducing any overhead if it is ignored.

chamikaramj replied:
I think the question to ask is what information will be more valuable for a Firestore user.
The pattern is:

"Generate data to write" -> "Write to firestore" -> "Do something else with the result"

I think most users will find failed results more useful but continuation when there are no failures is also important. 

I don't think you should worry about performance too much in either case. Overhead when the returned PCollection is ignored should be minimum for all major runners.

BenWhitehead replied:
I've refactored things to use the WriteSuccessSummary approach and pushed the commit.

chamikaramj commented on sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/firestore/FirestoreV1.java:
```diff
@@ -0,0 +1,620 @@
+ * <pre>{@code
+ * PCollection<Write> writes = ...;
+ * PDone sink = writes
+ *     .apply(FirestoreIO.v1().write().batchWrite().build());
```
Please make sure that these examples show any required parameters.

chamikaramj commented on sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/firestore/FirestoreV1.java:
```diff
@@ -0,0 +1,620 @@
+ * Alternatively, if you'd rather output write failures to a Dead Letter Queue add
+ * {@link BatchWrite.Builder#withDeadLetterQueue() withDeadLetterQueue} when building your writer.
+ * <pre>{@code
+ * PCollection<Write> writes = ...;
```
One option might be to return a WriteResult with getters to get a regular result PCollection or a dead-letter queue. We have something like that for BigQuery sink.

https://github.com/apache/beam/blob/0e47ea3a67acbb41b8f9357d722bbd565bc74105/sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/BigQueryIO.java#L1695

BenWhitehead replied:
Looking at https://github.com/apache/beam/blob/master/sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/WriteResult.java#L79-L107 it seems that the PCollections off of WriteResult are mutually exclusive and dependent on how it was constructed. I believe the builders I have accomplish the same functionality and are typed allowing for a compilation validation rather than a runtime validation.

chamikaramj replied:
Note that there WriteResult is an output (not that we return a PCollection<WriteResult>).
I'm not saying you should do that it was just an example.
See here for a previous discussion on this and various options.
https://lists.apache.org/thread.html/d1a4556a1e13a661cce19021926a5d0997fbbfde016d36989cf75a07%40%3Cdev.beam.apache.org%3E

chamikaramj commented on sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/firestore/FirestoreV1.java:
```diff
@@ -0,0 +1,620 @@
+
+    @Override
+    public PDone expand(PCollection<com.google.firestore.v1.Write> input) {
+      input.apply("batchWrite", ParDo.of(new DefaultBatchWriteFn(clock, firestoreStatefulComponentFactory, rpcQosOptions, CounterFactory.DEFAULT)));
```
Are writes to Firestore idempotent ? Note that bundles of this write step may fail and may be retried by the runner. So if the writes are not idempotent you'll have to handle this to prevent duplicate data from being written.

BenWhitehead replied:
Writes to firestore are generally idempotent, there are a few narrow cases where a precondition can be specified on an individual write (something like perform and update as long as the updatedAt matches timeX). If a precondition fails, firestire will respond with a failure status for that individual write, and if the dead letter queue is used will be output as an individual `WriteFailure{write, writeResult, status}`

chamikaramj commented on sdks/java/io/google-cloud-platform/src/test/java/org/apache/beam/sdk/io/gcp/firestore/it/BaseFirestoreIT.java:
```diff
@@ -0,0 +1,127 @@
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
```
It should be possible to trigger this IT from the PR to make sure it passes. Lemme know if you need help with this.

BenWhitehead replied:
I'm not sure how to trigger the ITs for the PR. They do need to run against a project that has a Firestore Native Mode database, otherwise they will be skipped (currently Firestore Native Mode and Datastore are mutually exclusive in a project).

chamikaramj reviewed (commented):
Thanks!

Seems like pre-commit and spotless failures are related. So please address them as well.

BenWhitehead commented on buildSrc/src/main/groovy/org/apache/beam/gradle/BeamModulePlugin.groovy:
```diff
@@ -425,6 +420,7 @@ class BeamModulePlugin implements Plugin<Project> {
     def errorprone_version = "2.3.4"
     def google_clients_version = "1.31.0"
     def google_cloud_bigdataoss_version = "2.1.6"
+    def google_cloud_firestore_version = "2.2.4"
```
I'll test and see if things work with the version provided by the BOM and report back.

BenWhitehead replied:
The version included in the bom for this base branch gets us firestore version 2.1.0 which is sufficient for this PR, so I've removed this version declaration and updated the rest of the gradle config to use the version from the BOM.

BenWhitehead commented:
I've fixed the spotless and firetore related failures in "Run Java PreCommit".

pabloem commented:
reopening to rerun tests

suztomo commented:
In another PR, I upgraded the version of the libraries-bom to 20.0.0 which has com.google.cloud:google-cloud-firestore:2.2.5. This may affect this PR.

BenWhitehead commented:
Thanks for the heads up @suztomo It shouldn't have a material impact to this PR. My changes to the gradle config are using the version from libraries-bom already, so I'd expect things to merge pretty easily and I can do that when I rebase and squash things before merging.

chamikaramj commented:
Sorry about the delay here. Lemme know if this is ready for another look.

chamikaramj commented:
Run Java PreCommit

chamikaramj commented:
Run Java_Examples_Dataflow PreCommit

BenWhitehead commented:
@chamikaramj Any preference related to this comment? https://github.com/apache/beam/pull/14261#discussion_r605893374

Other than that, someone from the Firestore Backend team should be taking a look and posting a review soon.

cynthiachi reviewed (changes requested):
No real changes, just a question about the 555 implementation, otherwise looks good!

danthev commented on sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/firestore/FirestoreV1WriteFn.java:
```diff
@@ -94,106 +108,124 @@ void handleWriteFailures(ContextAdapter<WriteFailure> context, Instant timestamp
-      builder
-          .include("rpcQosOptions", rpcQosOptions);
+    public final void populateDisplayData(
+        @edu.umd.cs.findbugs.annotations.NonNull DisplayData.Builder builder) {
```
Not checkerframework's `NonNull`?

BenWhitehead replied:
Left over codegen, will cleanup.

danthev commented on sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/firestore/RpcQosImpl.java:
```diff
@@ -204,7 +210,8 @@ public boolean awaitSafeToProceed(Instant instant) throws InterruptedException {
       if (shouldThrottleRequest.compareTo(Duration.ZERO) > 0) {
-        logger.info("Delaying request by {}ms", shouldThrottleRequest.getMillis());
+        long throttleRequestMillis = shouldThrottleRequest.getMillis();
+        logger.debug("Delaying request by {}ms", throttleRequestMillis);
```
This is when the adaptive throttler backs off, right? I think I'd tend toward an `INFO` log here, and a `due to previous failures` in the message.

BenWhitehead replied:
I dropped it to debug after adding the diagnostic distribution `qos_adaptiveThrottler_throttlingMs` so as not to blow out logs as this scenario can happen thousands of times for a job.

This screenshot shows the average number of times the adaptive throttler throttled grouped by docCount and fieldCount.

![image](https://user-images.githubusercontent.com/1754037/117376687-ad922000-ae9f-11eb-814f-95263cacd598.png)

danthev replied:
Is `throttlingMs` always reported, or does it hinge on that flag? As long as a user can easily find out what's happening that's fine.

danthev commented on sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/firestore/RpcQosImpl.java:
```diff
@@ -344,7 +354,8 @@ public boolean awaitSafeToProceed(Instant instant) throws InterruptedException {
         Duration throttleDuration = shouldThrottle.get();
-        getLogger().debug("Still ramping up, Delaying request by {}ms", throttleDuration.getMillis());
+        long throttleDurationMillis = throttleDuration.getMillis();
+        getLogger().debug("Still ramping up, Delaying request by {}ms", throttleDurationMillis);
```
Same here with `INFO` logs. If there's a better way that's fine too, what's important is that it's not too hard for the user to find out if they're being throttled by ramp-up or the adaptive throttler.

BenWhitehead replied:
Similar to adaptive throttler, this can happen a lot

![image](https://user-images.githubusercontent.com/1754037/117376940-27c2a480-aea0-11eb-803d-082dcdfc2f1f.png)

danthev commented on sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/firestore/RpcQosImpl.java:
```diff
@@ -83,48 +81,60 @@
+            filteringDistributionFactory);
+    writeRampUp =
+        new WriteRampUp(
+            Math.max(1, 500 / options.getHintMaxNumWorkers()), filteringDistributionFactory);
```
You can leave this if you want, but if it's not too much effort you could make the base budget a float and move `Math.max(1, ...)` into the per-second budget calculation. Then ramp-up isn't automatically faster for `hintMaxNumWorkers > 500`, but instead stays at the minimum level a little longer before actually ramping up.

BenWhitehead replied:
I'll push a commit with this change

BenWhitehead commented on sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/firestore/RpcQosImpl.java:
```diff
@@ -0,0 +1,909 @@
+   * Determines batch sizes based on past performance.
+   *
+   * <p>It aims for a target response time per RPC: it uses the response times for previous RPCs and
+   * the number of documents contained in them, calculates a rolling average time-per-document, and
```
Correct, right now it is only tracking number of documents not the size of those documents. It could be extended if we felt so inclined to do so at some point.

BenWhitehead commented on sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/firestore/RpcQosImpl.java:
```diff
@@ -0,0 +1,909 @@
+          distributionFactory.get(RpcQos.class.getName(), "qos_rampUp_availableWriteCountBudget");
+    }
+
+    int getAvailableWriteCountBudget(Instant instant) {
```
Correct, this only tracks data for up to one second. In the case of this ramp up calculation we're not taking past performance into consideration (instead the write batcher and adaptive throttler take care of that). This class is only attempt to limit based on the 555 budget and what has been used within that specific one-second window.

The budget value grows relative to the `firstInstant` captured, so even if a worker is only sending a request every 5 seconds the budget calculation will still be for the next time window. If the worker is sending more than one request a second, then it will decrement from the budget and be throttled until the next second if budget has been exhausted for the specific second.

For `hintMaxNumWorkers = 1` the budget will fill along this [line](https://www.wolframalpha.com/input/?i=500+*+1.5%5Emax%280%2C+%28x-5%29%2F5%29+from+0+to+30).

I've got a test `org.apache.beam.sdk.io.gcp.firestore.RpcQosSimulationTest#writeRampUp_shouldScaleAlongTheExpectedLine` which runs a series of operations against the RampUp to ensure values fall on the expected line.

BenWhitehead replied:
This is a screenshot of a job that ran for ~90min where we can see the growth steps for the the ramp up budget (the teal line/square line). Since this is aggregated across all workers it's not as exact as each of the 5 minute windows a single worker would see, but does show the progression.

![qos batch size calculation](https://user-images.githubusercontent.com/1754037/117369481-f7283e00-ae92-11eb-9ed1-b26caf51b27d.png)

For the same job we can see the actual (aggregated) min/mean/max batchCapacityCount over the duration of the job. We can see the steps pretty clearly on the left before capping out at 500, while at the same time the ramp up budget keeps increasing all the way up to almost 20k by the end of the job.

![qos_write_batchCapacityCount distribution](https://user-images.githubusercontent.com/1754037/117369946-a6fdab80-ae93-11eb-8c1c-15376551a11d.png)

BenWhitehead commented on sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/firestore/RpcQosImpl.java:
```diff
@@ -0,0 +1,909 @@
+      return budget;
+    }
+
+    // 500 * 1.5^max(0, (x-5)/5)
```
Yeah, I'll clean that up.

BenWhitehead commented:
@chamikaramj Both SRE and the Backend team have approved the changes, I've removed the `PDone` from the non-failure write type and fixed all related build failures. I think this is ready for the final review before I rebase it against master and prepare for merge.

chamikaramj commented:
Run Java PreCommit

chamikaramj commented:
Run SQL PreCommit

chamikaramj commented:
Run Java PreCommit

chamikaramj commented:
Run Java_Examples_Dataflow PreCommit

chamikaramj commented:
Run Python PreCommit

chamikaramj commented:
Run Java PostCommit

chamikaramj commented:
Run Java_Examples_Dataflow PreCommit

chamikaramj commented:
Run Python PreCommit

chamikaramj reviewed (approved):
Thanks. This looks great. We can merge after tests pass.

LGTM

BenWhitehead commented:
Awesome, thanks @chamikaramj! I'll squash all the commits and cleanup the commit message today.

BenWhitehead commented:
Failures in `Java ("Run Java PreCommit")` seem unrelated to my change.

All of my commits have been cleaned up and are marked with the jira issue for the feature and should be ready to merge.

chamikaramj commented:
Run Java PreCommit

chamikaramj commented:
Run Java PostCommit

chamikaramj commented:
Run Java PreCommit

chamikaramj commented:
BTW did you confirm that new ITs get captured by "Run Java PostCommit" ?

chamikaramj commented:
Don't see the new tests here: https://ci-beam.apache.org/job/beam_PostCommit_Java_PR/710/testReport/

chamikaramj commented:
Run PostCommit_Java_Dataflow

BenWhitehead commented:
I've dug into the missing integration tests running. The integration tests are currently using junit assumptions to skip running if the client isn't able to bootstrap and unfortunately this fact isn't being reported in the tests runs seemingly due to https://github.com/gradle/gradle/issues/5511 .

I'll prepare a separate PR with a fix to not use assumptions and update the gradle config to run Firestore tests against a project which has a Firestore native mode database present.

chamikaramj commented:
Run Java PreCommit

willbattel commented:
@BenWhitehead what was the rationale for using the `BatchWriter` over the new `BulkWriter` API (https://github.com/googleapis/java-firestore/pull/323)? The new `BulkWriter` was recommended by the Firestore team in a previous Beam PR (https://github.com/apache/beam/pull/10187#issuecomment-582216769), so I'm curious if this was taken into account or forgotten about. It would seem to me that the `BulkWriter` would be more appropriate for usage as a sink, compared to the added overhead and error frequency of `BatchWriter` for no apparent gain in this case. Or maybe I'm wrong, or there was another conversation about this that I'm not privy to.

BenWhitehead commented:
@willbattel After digging into the implementation of the Firestore connector, it became apparent that the Firestore SDK's model wouldn't be a good fit for use in the connector. (None of the model is Serializable, and there is no separation between data and operations which would have made it difficult to update the Firestore SDK to be beam compatible.) So here we've opted to use the gRPC client itself and forgo the Firestore SDK.

The [`com.google.cloud.firestore.BulkWriter`](https://github.com/googleapis/java-firestore/blob/master/google-cloud-firestore/src/main/java/com/google/cloud/firestore/Firestore.java#L181) from the Firestore SDK ultimately uses the same gRPC api [`google.firestore.v1.Firestore.BatchWrite`](https://cloud.google.com/firestore/docs/reference/rpc/google.firestore.v1#google.firestore.v1.Firestore.BatchWrite) which is being used in the connector here. The Firestore SDK has a `WriteBatch` feature which uses the [`google.firestore.v1.Firestore.Commit`](https://cloud.google.com/firestore/docs/reference/rpc/google.firestore.v1#google.firestore.v1.Firestore.Commit) api rather than `BatchWrite`.

Additionally the SRE team in particular wanted to build more client side traffic controls into the beam connector to ensure it is well behaved, since beam is so easy to scale up really fast. So in this sink, we have the 500/50/5 ramp up budgeting, adaptive throttling in conjunction with dynamic throughput base batch sizing which are not features which are supported by the Firestore SDK.

Secondarily, the `BulkWriter` from Firestore SDK doesn't currently provide doesn't provide any request lifecycle awareness hooks to allow for the `Metrics` which are surfaced by this sink.
